### PR TITLE
Replace yiq contrast by wcag official algorithm

### PR DESCRIPTION
**Pull request by Lausselloic**
Following @mdo PR for color update : https://github.com/twbs/bootstrap/pull/22836

Many thanks for the color contrast mixin which check the font color according to the background for badges only at time. But yiq is not the official wcag and doesn't give the same result on some colors (e.g with blue #527edb yiq set text color to white where official wcag lum set it to black.

I propose a SCSS adaptation of the HTMLCS JS solution. Need to add a dependency to sass-math-pow.

I hope this solution will be helpfull. Let me know

### CODE REVIEW

**Comment by Johann-S** [comment]
you should update our shrinkwrap because you added a new dependency with `npm run maintenance-shrinkwrap`

**Comment by Lausselloic** [comment]
Yes I've seen it. I will update till 1hour

File: scss/_functions.scss [hunk 1]
```diff
@@ -53,13 +56,28 @@
   $r: red($color);
   $g: green($color);
   $b: blue($color);
+  // get the relative lum for each color
+  $luminositecouleur: relativelum($r / 255, $g / 255, $b / 255);
```
Comment (patrickhlauke) [inline scss/_functions.scss]: small request: as the official language for our files is english, could you update the variables to `$luminositycolor` and `$luminositywhite` ? merci :)

Comment (Lausselloic) [inline scss/_functions.scss]: yes

Comment (patrickhlauke) [inline scss/_functions.scss]: actually, to stay consistent with current WCAG wording (though from what I remember there's discussion about it being incorrect), this should be `$luminancecolor` and `$luminancewhite`

Comment (patrickhlauke) [inline scss/_functions.scss]: sorry, you were quicker than my brain which needed more coffee ;)

**Comment by Lausselloic** [comment]
No problem I've many coffee :-)

**Comment by patrickhlauke** [comment]
LGTM but I'm not a SASS expert. @mdo et al, feel free to merge if this looks kosher to you.

**Comment by pat270** [comment]
Would be nice to have `@mixin color-yiq()` be a function that returns a value instead of printing a css declaration so we could use it in Sass variables.

**Comment by Lausselloic** [comment]
yes, that's my final idea, evolve that mixin to create a generic one that return a comparison between two colors. Or add in parameter output colors

**Comment by mdo** [comment]
> Need to add a dependency to sass-math-pow.

Not a fan of the dependency here. Is there another solution we could try? And if this changes what is output on our components, we'll need to revisit more things.

**Comment by patrickhlauke** [comment]
> if this changes what is output on our components, we'll need to revisit more things

I believe functionally the output is the same, just that it flips between light and dark text at a better calculated threshold which matches up with WCAG 2's requirements.

**Comment by mdo** [comment]
Yeah, I mean visually as much as the compiled code :).

**Comment by patrickhlauke** [comment]
Visually it'll be the same, except at the point just on the cusp between where light or dark would be better. There, it'll flip at a slightly earlier/later point, so yes for certain background colors it will be visually different.

**Comment by Lausselloic** [comment]
I understand and agree that adding a sass dependency is not fun. For @mdo the solution could be to Copy/paste source code directly in bootstrap utilities, because I don't think that pow expression will change in the time.
For visual consistence, would be possible to add optionnal input like : 
color-yiq($color, $chalenging-color-list, $light-fallback, $dark-fallback) {
Where $color will be the font or background color to check against an ordered color in $chalenging-color-list, and if none are valid then return light or dark fallback ?

for each $color in $chalenging-color-list -> check if ratio is valid, return the first valid color
if none have been return, check if light or dark fallback is better

**Comment by razh** [comment]
One alternative to the extra dependency would be to use the experimental [`functions`](https://github.com/sass/node-sass#functions--v300---experimental) feature in node-sass.

```js
const sass = require('sass')

const functions = {
  pow(base, exponent) {
    const result = Math.pow(base.getValue(), exponent.getValue())
    return new sass.types.Number(result)
  }
}
```

Would need some additional documentation for ruby-sass and other variants.

File: build/sass-functions.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
+'use strict'
+
+// Export math.pow function for node-sass
+
+var types = require('node-sass').types
```
Comment (Johann-S) [inline build/sass-functions.js]: you should use `const` instead of `var` here because you don't change `types`

**Comment by Lausselloic** [comment]
Hi, any news about this PR? I can update it if needed

**Comment by pierre-hilt** [comment]
Hello, do you have news about this issue? We would like to use WCAG contrast for color-yiq function. We tried to play with the treshold, but as it ia not the same function we can't have something perfect.

**Comment by Lausselloic** [comment]
Hello,

I use it in Boosted (that's why I create this PR) But wonder that one will go into the core soon.

**Comment by XhmikosR** [comment]
@Lausselloic: can you fetch upstream and rebase?

**Comment by MartijnCuppens** [comment]
I'm afraid we can't do this. This will make bootstrap not buildable with anything else than `node-sass` and it will require everyone to adapt their build with this custom config.

**Pull request closed without merging**